src/v/archival/archiver_service.cc

```diff
@@ -0,0 +1,999 @@
+    // during
+    // normal operation (note: ntp_archiver currently handles all errors
+    // internally)
+    archiver_failure,
+    // Archiver's async shutdown is completed
+    archiver_stopped,
+    // Shutdown initiated
+    shutdown,
+};
+
+std::ostream& operator<<(std::ostream& o, managed_partition_event_t e) {
+    switch (e) {
+    case managed_partition_event_t::leadership_acquired:
+        return o << "leadership_acquired";
+    case managed_partition_event_t::leadership_lost:
+        return o << "leadership_lost";
+    case managed_partition_event_t::archiver_started:
+        return o << "archiver_started";
+    case managed_partition_event_t::archiver_failure:
+        return o << "archiver_failure";
+    case managed_partition_event_t::archiver_stopped:
+        return o << "archiver_stopped";
+    case managed_partition_event_t::shutdown:
+        return o << "shutdown";
+    }
+}
+
+template<managed_partition_event_t event, class Derived>
+struct managed_partition_event_base : ssx::auto_fmt<Derived> {};
+
+/// Automagically prints any event object derived from
+/// managed_partition_event_base
+template<managed_partition_event_t event, class Derived>
+std::ostream& operator<<(
+  std::ostream& o, const managed_partition_event_base<event, Derived>& e) {
+    fmt::print(
+      o, "[{}..{}]", event, static_cast<const ssx::auto_fmt<Derived>>(e));
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const ss::future<>& f) {
+    fmt::print(o, "F<available?{},failed?{}>", f.available(), f.failed());
+    return o;
+}
+
+std::ostream&
+operator<<(std::ostream& o, const ss::lw_shared_ptr<ntp_archiver>&) {
+    return o;
+}
+
+// The 'managed_partition' is an STM with 4 states:
+enum class managed_partition_state_t {
+    //  _managed_ state means that the partition exists on a shard
+    //  but not necessary a leader. In this state we need to track
+    //  some information about the partition. This is an inactive
+    //  state. Object in this state can be disposed or transition to
+    //  _creating_ state when the node/shard becomes a leader. When
+    //  the state transitions to the _managed_ state we have to check
+    //  raft group state. If raft group is a leader we need to transition
+    //  to the _creating_ state.
+    managed,
+
+    //  _creating_ is a state in which the archiver is created. The
+    //  archiver is created/started asynchronously. From this state
+    //  we can transition to the _active_ state when the archiver is
+    //  created or we can transition to the _stopping_ state if the
+    //  leadership was lost while archiver was created.
+    creating,
+
+    //  _active_ state means that we're managing active archiver
+    //  instance which uploads data to the cloud storage asynchronously.
+    //  If the leadership is lost the state transitions to _stopping_.
+    //  The state is entered we need to check the status of the raft
```

Comment:
Don't comprehend this statement. Looks like you have a concern but I don't get it.